roachtest: introduce chaos, use in scaledata test

tbg opened the pull request:
I was tempted to go on an extended refactoring streak (pull out test and
log into their own packages, untangle dependencies, avoid relying on
roachprod monitor, ...) but I decided to approach this in a
goal-oriented manner.

Instead, code like this can now be introduced into existing tests
(assuming the workloads tolerate errors):

```sql
m.Go(randomKillChaos(
    c, m, chaosTimer(ctxWithTimeout, time.Minute, 30*time.Second), roachNodes))
```

Fixes #24279.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25427)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---

*[pkg/cmd/roachtest/chaos.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgXfR2FAaIUbYI-Yo:-LCFgXfR2FAaIUbYI-Yp:bb6k8mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L53)):*
> ```Go
> 			return err
> 		}
> 		for ctx.Err() == nil {
> ```

You're checking `ctx.Err()` here and in `chaosTimer` and `ctx.Done()` in the `select` below. This seems like overkill. Isn't the check in the `select` sufficient?

---

*[pkg/cmd/roachtest/chaos.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgw_i7s4DJES3BUCy:-LCFgw_i7s4DJES3BUCz:b-f2iwik) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L64)):*
> ```Go
> 			}
> 
> 			target := nodes[rand.Intn(len(nodes))]
> ```

Nit: perhaps better to maintain the `nodeListOption` abstraction and add a `nodeListOption.randNode()` method.

---

*[pkg/cmd/roachtest/cluster.go, line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhOjf5UwOa5c1cfZR:-LCFhOjf5UwOa5c1cfZS:bpeo84d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/cluster.go#L809)):*
> ```Go
> // ExpectDeath lets the monitor know that a node is about to be killed, and that
> // this should be ignored.
> func (m *monitor) ExpectDeath() {
> ```

Should we indicate which node we expect to die? Or is that too fragile? Rather than atomic, I think this could be implemented via a mutex and either a slice of node IDs are a set of node IDs.

---

*[pkg/cmd/roachtest/scaledata.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhsPiDBgw4pJyRWo-:-LCFhsPiDBgw4pJyRWo0:b-tg63uu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/scaledata.go#L99)):*
> ```Go
> 		// Kill one node at a time, with a minute of healthy cluster and thirty
> 		// seconds of down node.
> 		m.Go(randomKillChaos(c, m, chaosTimer(chaosCtx, time.Minute, 30*time.Second), roachNodes))
> ```

This is nice and concise!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Updated, PTAL. Once this is LGTSO (lgtm to someone) I'll run the scaledata tests on GCE a few times to make sure there's no funny business.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/chaos.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgXfR2FAaIUbYI-Yo:-LCH4sVu3rRjDR97awLQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're checking `ctx.Err()` here and in `chaosTimer` and `ctx.Done()` in the `select` below. This seems like overkill. Isn't the check in the `select` sufficient?
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/chaos.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgw_i7s4DJES3BUCy:-LCH4cWFCR5-ZWRlpcJz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L64)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: perhaps better to maintain the `nodeListOption` abstraction and add a `nodeListOption.randNode()` method.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/cluster.go, line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhOjf5UwOa5c1cfZR:-LCH4tfP8021_QScVMax:b3hdwj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/cluster.go#L809)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should we indicate which node we expect to die? Or is that too fragile? Rather than atomic, I think this could be implemented via a mutex and either a slice of node IDs are a set of node IDs.
</blockquote></details>

That was my initial approach, but when I went around to try it I didn't really see the point. If we expect a death, we definitely get a death (since we're going to kill something). If another node dies randomly, then it's either the one we're about to kill (and that wouldn't be prevented either way) or it's an unrelated node, so the counter will fire. I'd keep this KISS until we have a specific reason not to.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-:-LCH5L-LC3fEUG1aQYVN:bq9p4hw)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/chaos.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgXfR2FAaIUbYI-Yo:-LCK1be60GyZyQTCPAVM:b1npydt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L53)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This is better, though I'm still not clear on why `chaosTimer` needs to inspect `ctx.Err()`. And there is a race here in what is returned. If the context is done before `timing()` is invoked, you'll return `nil`. While if context becomes done between `timing()` and the `select`, then `ctx.Err()` is returned. 

Update: Oh, I just noticed that the `ctx` here and the one in `chaosTimer` are different. That's confusing, but this structure makes more sense now. Another approach would have been to pass in the total chaos duration to `randomKillChaos` and derive a new context from `ctx` using that timeout. Not sure if that is any better and it might confuse the error returned, so feel free to ignore this comment.

---

*[pkg/cmd/roachtest/cluster.go, line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhOjf5UwOa5c1cfZR:-LCK25YP9q2g1TS4id-p:bkrelnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/cluster.go#L809)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

That was my initial approach, but when I went around to try it I didn't really see the point. If we expect a death, we definitely get a death (since we're going to kill something). If another node dies randomly, then it's either the one we're about to kill (and that wouldn't be prevented either way) or it's an unrelated node, so the counter will fire. I'd keep this KISS until we have a specific reason not to.
</blockquote></details>

Ok, SGTM.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-:-LCK36tA2f6cfR5Up1kR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PTAL. I tested this revision with `./bin/roachtest run scaledata/filesystem_simulator/nodes=6` and it passed.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/chaos.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFgXfR2FAaIUbYI-Yo:-LCOwqfN4lcelxW3MECd:bg6y7c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/chaos.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is better, though I'm still not clear on why `chaosTimer` needs to inspect `ctx.Err()`. And there is a race here in what is returned. If the context is done before `timing()` is invoked, you'll return `nil`. While if context becomes done between `timing()` and the `select`, then `ctx.Err()` is returned. 

Update: Oh, I just noticed that the `ctx` here and the one in `chaosTimer` are different. That's confusing, but this structure makes more sense now. Another approach would have been to pass in the total chaos duration to `randomKillChaos` and derive a new context from `ctx` using that timeout. Not sure if that is any better and it might confuse the error returned, so feel free to ignore this comment.
</blockquote></details>

Take another look, I reworked the API a bit and in particular the context now doesn't matter.

---

*[pkg/cmd/roachtest/scaledata.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhsPiDBgw4pJyRWo-:-LCOwuxF4FQFWopRy52T:bgfqyaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/scaledata.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is nice and concise!
</blockquote></details>

It's maybe less concise now, but I think it's become even clearer (and still pretty concise!)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-:-LCOx-km91oHCMtHFEi6:btlsbiu)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/cmd/roachtest/scaledata.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhsPiDBgw4pJyRWo-:-LCPKPyuE6WR717GUEeC:b8p6ogq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/scaledata.go#L99)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's maybe less concise now, but I think it's become even clearer (and still pretty concise!)
</blockquote></details>

Still relatively concise. A slightly tweak on this approach would be to have the `Runner` function be global and to pass it a config struct. Something like:

```
m.Go(ChaosRunner(c, m, ChaosConfig{
  Timer: ...,
  Target: ...,
  Duration: ...,
})
```

What you currently have works fine too. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-:-LCPKpnJ3oZzwmL7pFdx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: introduce chaos, use in scaledata test

I was tempted to go on an extended refactoring streak (pull out test and
log into their own packages, untangle dependencies, avoid relying on
roachprod monitor, ...) but I decided to approach this in a
goal-oriented manner.

Instead, code like this can now be introduced into existing tests
(assuming the workloads tolerate errors):

```sql
m.Go(randomKillChaos(
    c, m, chaosTimer(ctxWithTimeout, time.Minute, 30*time.Second), roachNodes))
```

Fixes #24279.

Release note: None

tschottdorf commented:
bors r=petermattis

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cmd/roachtest/scaledata.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-LCFhsPiDBgw4pJyRWo-:-LCPLS2BEX0HQxf1SBDb:bvxf80t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1770ab9ba01c0dd5686da14fc24e6ca55adc8196/pkg/cmd/roachtest/scaledata.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Still relatively concise. A slightly tweak on this approach would be to have the `Runner` function be global and to pass it a config struct. Something like:

```
m.Go(ChaosRunner(c, m, ChaosConfig{
  Timer: ...,
  Target: ...,
  Duration: ...,
})
```

What you currently have works fine too. 
</blockquote></details>

I'll leave as is for now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25427#-:-LCPLWMS7F1R0VJDNRfU:blkd80i)*
<!-- Sent from Reviewable.io -->
